File: moonpoison/binary_tree-dfs/236.cpp
#include<bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
public:
    typedef TreeNode* T;
    TreeNode* lowestCommonAncestor(T root, T p, T q) 
    {
        vector<T> vp; vector<T> vq;
        vis(root, p, vp); vis(root, q, vq);
        for(int i=vq.size()-1;i>=0;--i)
            if(find(vp.begin(), vp.end(), vq[i])!=vp.end()) return vq[i];
        return NULL;
    }
    bool vis(T root, T p, vector<T> &v)
    {
        if(!root) return false;
        v.push_back(root);
        if(root==p) return true;
        if(vis(root->left, p, v)||vis(root->right, p, v)) return true;
        v.pop_back();
        return false;
    }
};

File: ice0208/binary-tree-dfs/236.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def go_down(self, cursor: TreeNode):
        p_f, q_f = False, False
        if cursor.val == self.p.val:
            p_f = True
        if cursor.val == self.q.val:
            q_f = True

        if cursor.left:
            p_f2, q_f2 = self.go_down(cursor.left)
            p_f = p_f or p_f2
            q_f = q_f or q_f2
            if p_f == q_f == True and self.answer == None:
                self.answer = cursor
        if cursor.right:
            p_f2, q_f2 = self.go_down(cursor.right)
            p_f = p_f or p_f2
            q_f = q_f or q_f2
            if p_f == q_f == True and self.answer == None:
                self.answer = cursor

        return (p_f, q_f)

    def lowestCommonAncestor(
        self, root: TreeNode, p: TreeNode, q: TreeNode
    ) -> TreeNode:
        self.p, self.q = p, q
        self.answer = None

        if root:
            self.go_down(root)

        return self.answer if self.answer else root
